File: backend/app/services/simulation_engine.py
"""
Industrial Simulation Engine
Generates realistic machine telemetry based on physics-based models.
"""

import asyncio
import random
import math
import time
from typing import Dict, Optional, Any
from dataclasses import dataclass, field


@dataclass
class MachineSimState:
    temperature: float = 25.0
    rpm: float = 0.0
    pressure: float = 1.0
    throughput: float = 0.0
    energy_consumption: float = 0.0
    efficiency: float = 90.0
    failure_probability: float = 2.0
    vibration: float = 2.0
    wear_level: float = 5.0
    material_flow: float = 0.0
    uptime: float = 0.0


class SimulationInstance:
    """A single running simulation with multiple machines."""

    def __init__(self, simulation_key: str, machines: list, tick_rate: int = 1000):
        self.key = simulation_key
        self.machines = machines
        self.tick_rate = tick_rate / 1000.0  # Convert to seconds
        self.running = False
        self.states: Dict[str, MachineSimState] = {}
        self.task: Optional[asyncio.Task] = None
        self._initialize_states()

    def _initialize_states(self):
        for machine in self.machines:
            machine_id = machine.get("id", f"machine_{random.randint(1000, 9999)}")
            self.states[machine_id] = MachineSimState(
                temperature=25 + random.uniform(0, 10),
                efficiency=85 + random.uniform(0, 10),
                failure_probability=random.uniform(1, 5),
                wear_level=random.uniform(5, 20),
            )

    async def start(self, broadcast_callback):
        """Start the simulation loop."""
        self.running = True
        self.task = asyncio.create_task(self._run_loop(broadcast_callback))

    async def stop(self):
        """Stop the simulation."""
        self.running = False
        if self.task:
            self.task.cancel()
            try:
                await self.task
            except asyncio.CancelledError:
                pass

    async def _run_loop(self, broadcast_callback):
        """Main simulation loop."""
        while self.running:
            telemetry_batch = {}

            for machine in self.machines:
                machine_id = machine.get("id")
                machine_type = machine.get("type", "generic")
                params = machine.get("parameters", {})

                state = self.states.get(machine_id)
                if not state:
                    continue

                # Simulate based on machine type
                updated_state = self._simulate(machine_type, params, state)
                self.states[machine_id] = updated_state

                # Build telemetry payload
                telemetry_batch[machine_id] = {
                    "machine_id": machine_id,
                    "timestamp": time.time(),
                    "temperature": round(updated_state.temperature, 2),
                    "rpm": round(updated_state.rpm, 1),
                    "pressure": round(updated_state.pressure, 3),
                    "throughput": round(updated_state.throughput, 2),
                    "energy_consumption": round(updated_state.energy_consumption, 2),
                    "machine_state": self._get_state_label(updated_state),
                    "failure_probability": round(updated_state.failure_probability, 1),
                    "maintenance_score": round(100 - updated_state.wear_level, 1),
                    "material_quantity": round(updated_state.material_flow, 2),
                    "efficiency_score": round(updated_state.efficiency, 1),
                    "sensor_health": round(95 + random.uniform(0, 5), 1),
                    "vibration": round(updated_state.vibration, 2),
                }

            # Broadcast to connected clients
            await broadcast_callback(self.key, telemetry_batch)
            await asyncio.sleep(self.tick_rate)

    def _simulate(self, machine_type: str, params: dict, state: MachineSimState) -> MachineSimState:
        """Route simulation to appropriate model."""
        state.uptime += self.tick_rate

        if machine_type == "battery_shredder":
            return self._sim_shredder(params, state)
        elif machine_type == "heated_chemical_tank":
            return self._sim_chemical_tank(params, state)
        elif machine_type == "filter_press":
            return self._sim_filter_press(params, state)
        elif machine_type == "steam_boiler":
            return self._sim_boiler(params, state)
        elif machine_type == "textile_dyeing_machine":
            return self._sim_dyeing(params, state)
        elif machine_type == "fabric_cutting_machine":
            return self._sim_cutting(params, state)
        else:
            return self._sim_generic(params, state)

    def _sim_shredder(self, params: dict, state: MachineSimState) -> MachineSimState:
        rpm = params.get("rpm", 5000)
        target_temp = 40 + (rpm / 10000) * 80

        state.temperature += (target_temp - state.temperature) * 0.1 + random.uniform(-1, 1)
        state.rpm = rpm + random.uniform(-50, 50)
        state.wear_level = min(100, state.wear_level + 0.01 + (0.05 if rpm > 7000 else 0))
        state.vibration = 2 + state.wear_level * 0.3 + random.uniform(-1, 1)
        state.efficiency = max(40, 95 - state.wear_level * 0.5)
        state.energy_consumption = 30 + (rpm / 10000) * 100 + random.uniform(-3, 3)
        state.throughput = params.get("shredding_rate", 800) * (state.efficiency / 100)
        state.material_flow = state.throughput
        state.pressure = 1 + random.uniform(0, 0.5)

        if rpm > 7000:
            state.failure_probability = min(95, state.failure_probability + 0.3)
        if state.temperature > 100:
            state.failure_probability = min(95, state.failure_probability + 0.5)

        return state

    def _sim_chemical_tank(self, params: dict, state: MachineSimState) -> MachineSimState:
        target_temp = params.get("tank_temperature", 85)
        pressure = params.get("pressure", 3)

        state.temperature += (target_temp - state.temperature) * 0.05 + random.uniform(-0.5, 0.5)
        state.pressure = pressure + random.uniform(-0.2, 0.2)
        state.rpm = params.get("mixing_speed", 120)

        temp_eff = 90 if 70 < state.temperature < 100 else 70
        state.efficiency = temp_eff + random.uniform(-3, 3)
        state.energy_consumption = params.get("heating_coil_power", 45) + random.uniform(-3, 3)
        state.throughput = params.get("tank_volume", 2000) * 0.01 * (state.efficiency / 100)
        state.material_flow = state.throughput
        state.vibration = 1 + random.uniform(0, 2)

        if state.temperature > 150:
            state.failure_probability = min(90, state.failure_probability + 1)

        return state

    def _sim_filter_press(self, params: dict, state: MachineSimState) -> MachineSimState:
        pressure = params.get("pressure_level", 8)

        state.pressure = pressure + random.uniform(-0.5, 0.5)
        state.temperature = 30 + random.uniform(-3, 3)
        state.wear_level = min(100, state.wear_level + 0.02)
        state.efficiency = max(50, 95 - state.wear_level * 0.4)
        state.energy_consumption = 15 + pressure * 2 + random.uniform(-2, 2)
        state.throughput = params.get("slurry_input", 500) * (state.efficiency / 100)
        state.material_flow = state.throughput * 0.85
        state.rpm = 0
        state.vibration = 1 + random.uniform(0, 1)

        if state.wear_level > 70:
            state.failure_probability = min(80, state.failure_probability + 0.3)

        return state

    def _sim_boiler(self, params: dict, state: MachineSimState) -> MachineSimState:
        target_pressure = params.get("steam_pressure", 6)
        safety_limit = params.get("safety_pressure", 12)

        state.pressure = target_pressure + random.uniform(-0.3, 0.3)
        state.temperature = 100 + state.pressure * 15 + random.uniform(-3, 3)
        state.energy_consumption = params.get("fuel_consumption", 30) * 0.8 + random.uniform(-3, 3)
        state.efficiency = 85 + random.uniform(-5, 5)
        state.throughput = params.get("heat_output", 200) * (state.efficiency / 100)
        state.rpm = 0
        state.vibration = 2 + random.uniform(0, 3)
        state.material_flow = state.throughput

        if state.pressure > safety_limit * 0.8:
            state.failure_probability = min(90, state.failure_probability + 0.5)

        return state

    def _sim_dyeing(self, params: dict, state: MachineSimState) -> MachineSimState:
        target_temp = params.get("dye_temperature", 80)

        state.temperature += (target_temp - state.temperature) * 0.08 + random.uniform(-1, 1)
        state.rpm = params.get("rotation_speed", 30) + random.uniform(-3, 3)
        state.pressure = 1 + random.uniform(0, 0.3)
        state.energy_consumption = params.get("heating_load", 35) + random.uniform(-2, 2)
        state.efficiency = 80 + random.uniform(-5, 5)
        state.throughput = 50 + state.efficiency * 0.5
        state.material_flow = state.throughput
        state.vibration = 1 + random.uniform(0, 2)

        if state.temperature > 120:
            state.failure_probability = min(70, state.failure_probability + 0.3)

        return state

    def _sim_cutting(self, params: dict, state: MachineSimState) -> MachineSimState:
        speed = params.get("cutting_speed", 15)

        state.temperature = 25 + speed * 1.5 + random.uniform(-2, 2)
        state.rpm = speed * 100 + random.uniform(-30, 30)
        state.wear_level = min(100, state.wear_level + 0.015)
        state.efficiency = max(60, 95 - state.wear_level * 0.3)
        state.energy_consumption = params.get("power_consumption", 8) + speed * 0.3 + random.uniform(-1, 1)
        state.throughput = speed * (state.efficiency / 100) * 10
        state.material_flow = state.throughput
        state.pressure = 1
        state.vibration = 2 + state.wear_level * 0.1 + random.uniform(-1, 1)

        return state

    def _sim_generic(self, params: dict, state: MachineSimState) -> MachineSimState:
        state.temperature = max(20, min(120, state.temperature + random.uniform(-2, 2)))
        state.rpm = 1000 + random.uniform(-100, 100)
        state.pressure = 1 + random.uniform(0, 2)
        state.efficiency = max(60, min(99, state.efficiency + random.uniform(-2, 2)))
        state.energy_consumption = 20 + random.uniform(-5, 5)
        state.throughput = 100 + random.uniform(-10, 10)
        state.material_flow = state.throughput
        state.vibration = 2 + random.uniform(0, 3)
        state.wear_level = min(100, state.wear_level + 0.01)
        return state

    def _get_state_label(self, state: MachineSimState) -> str:
        if state.failure_probability > 70:
            return "critical"
        if state.failure_probability > 40 or state.temperature > 100:
            return "warning"
        if state.efficiency < 50:
            return "degraded"
        return "running"

    def get_current_state(self) -> dict:
        """Get current state of all machines."""
        return {
            machine_id: {
                "temperature": state.temperature,
                "rpm": state.rpm,
                "pressure": state.pressure,
                "throughput": state.throughput,
                "energy_consumption": state.energy_consumption,
                "efficiency": state.efficiency,
                "failure_probability": state.failure_probability,
                "vibration": state.vibration,
                "wear_level": state.wear_level,
                "state": self._get_state_label(state),
            }
            for machine_id, state in self.states.items()
        }


class SimulationManager:
    """Manages multiple simulation instances."""

    def __init__(self):
        self.simulations: Dict[str, SimulationInstance] = {}

    def status(self) -> dict:
        return {
            "active_simulations": len([s for s in self.simulations.values() if s.running]),
            "total_simulations": len(self.simulations),
        }

    def create_simulation(self, key: str, machines: list, tick_rate: int = 1000) -> SimulationInstance:
        sim = SimulationInstance(key, machines, tick_rate)
        self.simulations[key] = sim
        return sim

    async def start_simulation(self, key: str, ws_manager) -> bool:
        sim = self.simulations.get(key)
        if not sim:
            return False

        async def broadcast(sim_key, telemetry_batch):
            await ws_manager.broadcast(sim_key, {
                "type": "telemetry",
                "data": telemetry_batch,
            })

        await sim.start(broadcast)
        return True

    async def stop_simulation(self, key: str) -> bool:
        sim = self.simulations.get(key)
        if not sim:
            return False
        await sim.stop()
        return True

    async def pause_simulation(self, key: str) -> bool:
        sim = self.simulations.get(key)
        if not sim:
            return False
        sim.running = False
        return True

    def get_state(self, key: str) -> dict:
        sim = self.simulations.get(key)
        if not sim:
            return {}
        return sim.get_current_state()

    async def stop_all(self):
        for sim in self.simulations.values():
            await sim.stop()

File: frontend/src/engine/SimulationEngine.ts
import { MachineConfig, TelemetryData, Alert, CustomIndustry, MachineDefinition } from '@/types'
import { ALL_MACHINES } from '@/data/machines'
import { findMachineDefinition } from '@/data/industryRegistry'

/**
 * Simulation engine that supports both built-in machine types (with bespoke
 * physics) and AI-generated types (via parameter-driven generic simulation).
 */
export class SimulationEngine {
  private machines: MachineConfig[] = []
  private interval: NodeJS.Timeout | null = null
  private tickRate: number = 1000
  private onTelemetry: (machineId: string, data: TelemetryData) => void
  private onAlert: (alert: Alert) => void
  private machineStates: Map<string, MachineSimState> = new Map()
  private customIndustries: CustomIndustry[]

  constructor(
    machines: MachineConfig[],
    tickRate: number,
    customIndustries: CustomIndustry[],
    onTelemetry: (machineId: string, data: TelemetryData) => void,
    onAlert: (alert: Alert) => void
  ) {
    this.machines = machines
    this.tickRate = tickRate
    this.customIndustries = customIndustries
    this.onTelemetry = onTelemetry
    this.onAlert = onAlert
    this.initializeStates()
  }

  private initializeStates() {
    this.machines.forEach((machine) => {
      this.machineStates.set(machine.id, {
        temperature: 25 + Math.random() * 10,
        rpm: 0,
        pressure: 1,
        throughput: 0,
        energyConsumption: 0,
        efficiency: 85 + Math.random() * 10,
        failureProbability: Math.random() * 5,
        vibration: Math.random() * 3,
        wearLevel: Math.random() * 20,
        materialFlow: 0,
        uptime: 0,
      })
    })
  }

  start() {
    this.interval = setInterval(() => this.tick(), this.tickRate)
  }

  stop() {
    if (this.interval) {
      clearInterval(this.interval)
      this.interval = null
    }
  }

  private tick() {
    this.machines.forEach((machine) => {
      const state = this.machineStates.get(machine.id)
      if (!state) return

      const updatedState = this.simulateMachine(machine, state)
      this.machineStates.set(machine.id, updatedState)

      const telemetry: TelemetryData = {
        machineId: machine.id,
        timestamp: Date.now(),
        temperature: updatedState.temperature,
        rpm: updatedState.rpm,
        pressure: updatedState.pressure,
        throughput: updatedState.throughput,
        energyConsumption: updatedState.energyConsumption,
        machineState: this.getMachineState(updatedState),
        failureProbability: updatedState.failureProbability,
        maintenanceScore: 100 - updatedState.wearLevel,
        materialQuantity: updatedState.materialFlow,
        efficiencyScore: updatedState.efficiency,
        sensorHealth: 95 + Math.random() * 5,
        vibration: updatedState.vibration,
      }

      this.onTelemetry(machine.id, telemetry)
      this.checkAlerts(machine, updatedState)
    })
  }

  private simulateMachine(machine: MachineConfig, state: MachineSimState): MachineSimState {
    const params = machine.parameters
    const newState = { ...state }
    newState.uptime += this.tickRate / 1000

    // Built-in physics first
    switch (machine.type) {
      case 'battery_shredder':
        return this.simulateShredder(params, newState)
      case 'heated_chemical_tank':
        return this.simulateChemicalTank(params, newState)
      case 'filter_press':
        return this.simulateFilterPress(params, newState)
      case 'steam_boiler':
        return this.simulateBoiler(params, newState)
      case 'textile_dyeing_machine':
        return this.simulateDyeingMachine(params, newState)
      case 'fabric_cutting_machine':
        return this.simulateCuttingMachine(params, newState)
    }

    // AI-generated or unknown — derive a plausible model from parameters
    return this.simulateParameterDriven(machine, newState)
  }

  /**
   * Derives temperature, rpm, pressure, throughput, and energy targets from
   * parameter names so AI-generated machines produce useful telemetry.
   */
  private simulateParameterDriven(
    machine: MachineConfig,
    state: MachineSimState
  ): MachineSimState {
    const def = findMachineDefinition(machine.type, this.customIndustries)
    const p = machine.parameters
    const has = (key: string) => p[key] !== undefined

    // Lookup helpers
    const get = (keys: string[], fallback: number) => {
      for (const k of keys) {
        if (p[k] !== undefined) return p[k]
        // fuzzy match
        const found = Object.keys(p).find((pk) => pk.includes(k))
        if (found) return p[found]
      }
      return fallback
    }

    // Temperature target
    const tempTarget = get(['temperature', 'tank_temperature', 'press_temperature', 'water_temperature'], 30)
    state.temperature += (tempTarget - state.temperature) * 0.08 + (Math.random() - 0.5) * 1.2

    // RPM
    const rpmTarget = get(['rpm', 'rotation_speed', 'drum_speed', 'mixing_speed', 'speed', 'cycle_rate'], 500)
    state.rpm = rpmTarget + (Math.random() - 0.5) * (rpmTarget * 0.05)

    // Pressure
    const pressureTarget = get(['pressure', 'pressure_level', 'steam_pressure', 'hydraulic_pressure'], 1)
    state.pressure = pressureTarget + (Math.random() - 0.5) * 0.4

    // Energy
    const energyTarget = get(['energy_usage', 'energy_consumption', 'power_consumption', 'heating_load', 'fuel_consumption'], 20)
    state.energyConsumption = energyTarget * (0.9 + Math.random() * 0.2)

    // Throughput target — try several
    const throughputTarget = get(['throughput', 'shredding_rate', 'feed_rate', 'intake_rate', 'cutting_speed', 'inspection_speed', 'packaging_rate'], 200)

    // Wear & efficiency progression
    state.wearLevel = Math.min(100, state.wearLevel + 0.012 + (state.rpm > rpmTarget * 1.2 ? 0.04 : 0))
    state.efficiency = Math.max(45, 95 - state.wearLevel * 0.4 + (Math.random() - 0.5) * 4)

    state.throughput = throughputTarget * (state.efficiency / 100)
    state.materialFlow = state.throughput

    // Vibration grows with wear, especially for rotating machines
    const rotating = state.rpm > 100
    state.vibration = (rotating ? 2 : 1) + state.wearLevel * (rotating ? 0.18 : 0.05) + (Math.random() - 0.5) * 1.2

    // Failure probability
    const overheating = state.temperature > tempTarget * 1.4 + 20
    const overpressure = state.pressure > pressureTarget * 1.5 + 1
    if (overheating) state.failureProbability = Math.min(95, state.failureProbability + 0.6)
    if (overpressure) state.failureProbability = Math.min(95, state.failureProbability + 0.4)
    if (state.wearLevel > 70) state.failureProbability = Math.min(95, state.failureProbability + 0.25)

    void def // currently unused but kept for future per-output customization
    return state
  }

  private simulateShredder(params: Record<string, number>, state: MachineSimState): MachineSimState {
    const rpm = params.rpm || 5000
    const targetTemp = 40 + (rpm / 10000) * 80

    state.temperature += (targetTemp - state.temperature) * 0.1 + (Math.random() - 0.5) * 2
    state.rpm = rpm + (Math.random() - 0.5) * 100

    state.wearLevel = Math.min(100, state.wearLevel + 0.01 + (rpm > 7000 ? 0.05 : 0))
    state.vibration = 2 + state.wearLevel * 0.3 + (Math.random() - 0.5) * 2
    state.efficiency = Math.max(40, 95 - state.wearLevel * 0.5)
    state.energyConsumption = 30 + (rpm / 10000) * 100 + Math.random() * 5
    state.throughput = (params.shredding_rate || 800) * (state.efficiency / 100)
    state.materialFlow = state.throughput

    if (rpm > 7000) state.failureProbability = Math.min(95, state.failureProbability + 0.5)
    if (state.temperature > 100) state.failureProbability = Math.min(95, state.failureProbability + 1)
    if (state.vibration > 20) state.failureProbability = Math.min(95, state.failureProbability + 0.3)

    state.pressure = 1 + Math.random() * 0.5
    return state
  }

  private simulateChemicalTank(params: Record<string, number>, state: MachineSimState): MachineSimState {
    const targetTemp = params.tank_temperature || 85
    const pressure = params.pressure || 3

    state.temperature += (targetTemp - state.temperature) * 0.05 + (Math.random() - 0.5) * 1
    state.pressure = pressure + (Math.random() - 0.5) * 0.3
    state.rpm = params.mixing_speed || 120

    const tempEfficiency = state.temperature > 70 && state.temperature < 100 ? 90 : 70
    state.efficiency = tempEfficiency + (Math.random() - 0.5) * 5

    state.energyConsumption = (params.heating_coil_power || 45) + Math.random() * 5
    state.throughput = (params.tank_volume || 2000) * 0.01 * (state.efficiency / 100)
    state.materialFlow = state.throughput

    if (state.temperature > 150) state.failureProbability = Math.min(90, state.failureProbability + 2)

    state.vibration = 1 + Math.random() * 2
    return state
  }

  private simulateFilterPress(params: Record<string, number>, state: MachineSimState): MachineSimState {
    const pressure = params.pressure_level || 8
    state.pressure = pressure + (Math.random() - 0.5) * 1
    state.temperature = 30 + Math.random() * 10
    state.wearLevel = Math.min(100, state.wearLevel + 0.02)
    state.efficiency = Math.max(50, 95 - state.wearLevel * 0.4)
    state.energyConsumption = 15 + pressure * 2 + Math.random() * 3
    state.throughput = (params.slurry_input || 500) * (state.efficiency / 100)
    state.materialFlow = state.throughput * 0.85
    state.rpm = 0
    state.vibration = 1 + Math.random()
    if (state.wearLevel > 70) state.failureProbability = Math.min(80, state.failureProbability + 0.5)
    return state
  }

  private simulateBoiler(params: Record<string, number>, state: MachineSimState): MachineSimState {
    const targetPressure = params.steam_pressure || 6
    const safetyLimit = params.safety_pressure || 12
    state.pressure = targetPressure + (Math.random() - 0.5) * 0.5
    state.temperature = 100 + state.pressure * 15 + (Math.random() - 0.5) * 5
    state.energyConsumption = (params.fuel_consumption || 30) * 0.8 + Math.random() * 5
    state.efficiency = 85 + (Math.random() - 0.5) * 10
    state.throughput = (params.heat_output || 200) * (state.efficiency / 100)
    state.rpm = 0
    state.vibration = 2 + Math.random() * 3
    state.materialFlow = state.throughput
    if (state.pressure > safetyLimit * 0.8) state.failureProbability = Math.min(90, state.failureProbability + 1)
    return state
  }

  private simulateDyeingMachine(params: Record<string, number>, state: MachineSimState): MachineSimState {
    const targetTemp = params.dye_temperature || 80
    state.temperature += (targetTemp - state.temperature) * 0.08 + (Math.random() - 0.5) * 1.5
    state.rpm = (params.rotation_speed || 30) + (Math.random() - 0.5) * 5
    state.pressure = 1 + Math.random() * 0.3
    state.energyConsumption = (params.heating_load || 35) + Math.random() * 3
    state.efficiency = 80 + (Math.random() - 0.5) * 10
    state.throughput = 50 + state.efficiency * 0.5
    state.materialFlow = state.throughput
    state.vibration = 1 + Math.random() * 2
    if (state.temperature > 120) state.failureProbability = Math.min(70, state.failureProbability + 0.5)
    return state
  }

  private simulateCuttingMachine(params: Record<string, number>, state: MachineSimState): MachineSimState {
    const speed = params.cutting_speed || 15
    state.temperature = 25 + speed * 1.5 + (Math.random() - 0.5) * 3
    state.rpm = speed * 100 + (Math.random() - 0.5) * 50
    state.wearLevel = Math.min(100, state.wearLevel + 0.015)
    state.efficiency = Math.max(60, 95 - state.wearLevel * 0.3)
    state.energyConsumption = (params.power_consumption || 8) + speed * 0.3 + Math.random() * 2
    state.throughput = speed * (state.efficiency / 100) * 10
    state.materialFlow = state.throughput
    state.pressure = 1
    state.vibration = 2 + state.wearLevel * 0.1 + Math.random() * 2
    return state
  }

  private getMachineState(state: MachineSimState): string {
    if (state.failureProbability > 70) return 'critical'
    if (state.failureProbability > 40 || state.temperature > 100) return 'warning'
    if (state.efficiency < 50) return 'degraded'
    return 'running'
  }

  private checkAlerts(machine: MachineConfig, state: MachineSimState) {
    const def: MachineDefinition | undefined =
      ALL_MACHINES.find((m) => m.type === machine.type) ||
      findMachineDefinition(machine.type, this.customIndustries)
    const machineName = def?.name || machine.name

    if (state.temperature > 120) {
      this.onAlert({
        id: `alert_${machine.id}_temp_${Date.now()}`,
        machineId: machine.id,
        machineName,
        type: 'critical',
        category: 'temperature',
        message: `Overheating detected: ${state.temperature.toFixed(1)}°C exceeds safe threshold`,
        timestamp: Date.now(),
        acknowledged: false,
      })
    } else if (state.temperature > 90) {
      this.onAlert({
        id: `alert_${machine.id}_temp_warn_${Date.now()}`,
        machineId: machine.id,
        machineName,
        type: 'warning',
        category: 'temperature',
        message: `High temperature: ${state.temperature.toFixed(1)}°C approaching limit`,
        timestamp: Date.now(),
        acknowledged: false,
      })
    }

    if (state.vibration > 25) {
      this.onAlert({
        id: `alert_${machine.id}_vib_${Date.now()}`,
        machineId: machine.id,
        machineName,
        type: 'critical',
        category: 'vibration',
        message: `Abnormal vibration: ${state.vibration.toFixed(1)} mm/s — possible mechanical failure`,
        timestamp: Date.now(),
        acknowledged: false,
      })
    }

    if (state.failureProbability > 60) {
      this.onAlert({
        id: `alert_${machine.id}_fail_${Date.now()}`,
        machineId: machine.id,
        machineName,
        type: 'critical',
        category: 'failure',
        message: `High failure probability: ${state.failureProbability.toFixed(0)}% — immediate maintenance recommended`,
        timestamp: Date.now(),
        acknowledged: false,
      })
    }

    if (state.wearLevel > 80) {
      this.onAlert({
        id: `alert_${machine.id}_maint_${Date.now()}`,
        machineId: machine.id,
        machineName,
        type: 'warning',
        category: 'maintenance',
        message: `Component wear at ${state.wearLevel.toFixed(0)}% — schedule maintenance`,
        timestamp: Date.now(),
        acknowledged: false,
      })
    }

    if (state.pressure > 10) {
      this.onAlert({
        id: `alert_${machine.id}_press_${Date.now()}`,
        machineId: machine.id,
        machineName,
        type: 'critical',
        category: 'pressure',
        message: `Pressure anomaly: ${state.pressure.toFixed(1)} bar exceeds safety threshold`,
        timestamp: Date.now(),
        acknowledged: false,
      })
    }
  }
}

interface MachineSimState {
  temperature: number
  rpm: number
  pressure: number
  throughput: number
  energyConsumption: number
  efficiency: number
  failureProbability: number
  vibration: number
  wearLevel: number
  materialFlow: number
  uptime: number
}
